File: golang/evaluate_reverse_polish_notation.go
package main

import "strconv"

// Space     : O(n)
// Time      : O(n)

func evalRPN(tokens []string) int {
	stack := []int{}
	n := len(tokens)

	for i := 0; i < n; i++ {
		if tokens[i] == "+" {
			res := stack[len(stack)-2] + stack[len(stack)-1]
			stack = stack[:len(stack)-2]
			stack = append(stack, res)
		} else if tokens[i] == "-" {
			res := stack[len(stack)-2] - stack[len(stack)-1]
			stack = stack[:len(stack)-2]
			stack = append(stack, res)
		} else if tokens[i] == "*" {
			res := stack[len(stack)-2] * stack[len(stack)-1]
			stack = stack[:len(stack)-2]
			stack = append(stack, res)
		} else if tokens[i] == "/" {
			res := stack[len(stack)-2] / stack[len(stack)-1]
			stack = stack[:len(stack)-2]
			stack = append(stack, res)
		} else {
			digit, _ := strconv.Atoi(tokens[i])
			stack = append(stack, digit)
		}
	}

	return stack[0]
}

File: python3/evaluate_reverse_polish_notation.py
# Space     : O(n)
# Time      : O(n)

class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        n = len(tokens)
        
        for i in range(n):
            if tokens[i] in ["+",  "-", "*", "/"]:
                num2, num1 = stack.pop(), stack.pop()
                if tokens[i] == "+":
                    stack.append(num1 + num2)
                elif tokens[i] == "-":
                    stack.append(num1 - num2)
                elif tokens[i] == "*":
                    stack.append(num1 * num2)
                else:
                    stack.append(int(num1 / num2))
            else:
                stack.append(int(tokens[i]))

        return stack[0]
        